presto-main/src/main/java/com/facebook/presto/operator/aggregation/noisyaggregation/SfmSketchAggregationUtils.java

```diff
@@ -0,0 +1,151 @@
+
+    private SfmSketchAggregationUtils() {}
+
+    public static void ensureStateInitialized(SfmSketchState state, double epsilon, long numberOfBuckets, long precision)
+    {
+        if (state.getSketch() != null) {
+            return; // already initialized
+        }
+
+        validateSketchParameters(epsilon, (int) numberOfBuckets, (int) precision);
+        SfmSketch sketch = SfmSketch.create((int) numberOfBuckets, (int) precision);
+        state.setSketch(sketch);
+        state.setEpsilon(epsilon);
+        state.addMemoryUsage(sketch.getRetainedSizeInBytes());
+    }
+
+    public static void validateSketchParameters(double epsilon, int numberOfBuckets, int precision)
+    {
+        checkCondition(epsilon > 0,
+                INVALID_FUNCTION_ARGUMENT, "epsilon must be positive");
+        checkCondition(numberOfBuckets > 0 && (numberOfBuckets & (numberOfBuckets - 1)) == 0,
+                INVALID_FUNCTION_ARGUMENT, "numberOfBuckets must be a power of 2");
+        checkCondition(precision > 0 && precision % Byte.SIZE == 0,
+                INVALID_FUNCTION_ARGUMENT, "precision must be a multiple of %s", Byte.SIZE);
+    }
+
+    public static void addHashToSketch(SfmSketchState state, long hash, double epsilon, long numberOfBuckets, long precision)
+    {
+        ensureStateInitialized(state, epsilon, numberOfBuckets, precision);
+        state.getSketch().addHash(hash);
+    }
+
+    public static long hashLong(MethodHandle methodHandle, long value)
+    {
+        long hash;
+        try {
+            hash = (long) methodHandle.invokeExact(value);
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+        return hash;
+    }
+
+    public static long hashDouble(MethodHandle methodHandle, double value)
+    {
+        long hash;
+        try {
+            hash = (long) methodHandle.invokeExact(value);
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+        return hash;
+    }
+
+    public static long hashSlice(MethodHandle methodHandle, Slice value)
+    {
+        long hash;
+        try {
+            hash = (long) methodHandle.invokeExact(value);
+        }
+        catch (Throwable t) {
+            throw internalError(t);
+        }
+        return hash;
+    }
+
+    public static void mergeStates(SfmSketchState state, SfmSketchState otherState)
+    {
+        SfmSketch sketch = state.getSketch();
+        SfmSketch otherSketch = otherState.getSketch();
+        if (sketch == null) {
+            state.setSketch(otherSketch);
+            state.setEpsilon(otherState.getEpsilon());
+            state.addMemoryUsage(otherSketch.getRetainedSizeInBytes());
+        }
+        else {
+            try {
+                // Throws if the sketches are incompatible (e.g., different bucket counts/size)
+                // Catch and throw a PrestoException
+                sketch.mergeWith(otherSketch);
```

Comment:
same memory accounting related question here

Reply:
As above, memory usage won't change here, but I'll add the accounting back for future-proof-ness.

Reply:
(This is no longer true and memory accounting has been adjusted.)